pkg/promtail/targets/syslogtarget.go

```diff
@@ -0,0 +1,323 @@
+
+	listener net.Listener
+	messages chan message
+
+	ctx             context.Context
+	ctxCancel       context.CancelFunc
+	openConnections *sync.WaitGroup
+}
+
+type message struct {
+	labels  model.LabelSet
+	message string
+}
+
+// NewSyslogTarget configures a new SyslogTarget.
+func NewSyslogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	config *scrape.SyslogTargetConfig,
+) (*SyslogTarget, error) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	t := &SyslogTarget{
+		logger:        logger,
+		handler:       handler,
+		config:        config,
+		relabelConfig: relabel,
+
+		ctx:             ctx,
+		ctxCancel:       cancel,
+		openConnections: new(sync.WaitGroup),
+	}
+
+	t.messages = make(chan message)
+	go t.messageSender()
+
+	err := t.run()
+	return t, err
+}
+
+func (t *SyslogTarget) run() error {
+	l, err := net.Listen("tcp", t.config.ListenAddress)
+	l = conntrack.NewListener(l, conntrack.TrackWithName("syslog_target/"+t.config.ListenAddress))
+	if err != nil {
+		return fmt.Errorf("error setting up syslog target %w", err)
+	}
+	t.listener = l
+	level.Info(t.logger).Log("msg", "syslog listening on address", "address", t.ListenAddress().String())
+
+	t.openConnections.Add(1)
+	go t.acceptConnections()
+
+	return nil
+}
+
+func (t *SyslogTarget) acceptConnections() {
+	defer t.openConnections.Done()
+
+	l := log.With(t.logger, "address", t.listener.Addr().String())
+
+	backoff := util.NewBackoff(t.ctx, util.BackoffConfig{
+		MinBackoff: 5 * time.Millisecond,
+		MaxBackoff: 1 * time.Second,
+	})
+
+	for {
+		c, err := t.listener.Accept()
+		if err != nil {
+			select {
+			case <-t.ctx.Done():
+				level.Info(l).Log("msg", "syslog server shutting down")
+				return
+			default:
+			}
+			if ne, ok := err.(net.Error); ok && ne.Temporary() {
+				level.Warn(l).Log("msg", "failed to accept syslog connection", "err", err, "num_retries", backoff.NumRetries())
+				backoff.Wait()
+				continue
+			}
+
+			level.Error(l).Log("msg", "failed to accept syslog connection. quiting", "err", err)
+			return
+		}
+		backoff.Reset()
+
+		t.openConnections.Add(1)
+		go t.handleConnection(c)
+	}
+}
+
+func (t *SyslogTarget) handleConnection(cn net.Conn) {
+	defer t.openConnections.Done()
+
+	c := &idleTimeoutConn{cn, t.idleTimeout()}
+
+	handlerCtx, cancel := context.WithCancel(t.ctx)
+	defer cancel()
+	go func() {
+		<-handlerCtx.Done()
+		_ = c.Close()
+	}()
+
+	connLabels := t.connectionLabels(c)
+
+	for msg := range syslogparser.ParseStream(c) {
+		if err := msg.Error; err != nil {
+			t.handleMessageError(err)
+			continue
+		}
+		t.handleMessage(connLabels.Copy(), msg.Message)
+	}
+}
+
+func (t *SyslogTarget) handleMessageError(err error) {
+	var ne net.Error
+	if errors.As(err, &ne) && ne.Timeout() {
+		level.Debug(t.logger).Log("msg", "connection timed out", "err", ne)
+		return
+	}
+	level.Debug(t.logger).Log("msg", "error parsing syslog stream", "err", err)
```

Comment:
This should be at least a warning, so that it's not hidden if someone isn't using debug log.